### PR TITLE
Feat/group by name

### PR DESCRIPTION
**Pull request by tonsV2**
This PR is mostly a lot of s/groupId/groupName/ but also eliminating of "global" docs.go.

Merge https://github.com/dhis2-sre/im-database-manager/pull/25 before this PR!

### CODE REVIEW

File: internal/handler/authorize.go [hunk 1]
```diff
@@ -1,52 +1,33 @@
...
-		return groups[i].ID <= groups[j].ID
-	})
-
-	index := sort.Search(len(groups), func(i int) bool {
-		return uint(groups[i].ID) >= groupId
-	})
-
-	return index < len(groups) && uint(groups[index].ID) == groupId
+func isMemberOf(groupName string, groups []*models.Group) bool {
+	f := func(g *models.Group) bool { return g.Name == groupName }
```
Comment (teleivo) [inline internal/handler/authorize.go:26]: Same as in im-user (I think that's where we have the same). Inline or replace with range without the need for dependency (I would prefer it that way).

File: internal/handler/authorize_test.go [hunk 2]
```diff
@@ -10,20 +10,20 @@ import (
 
 func TestCanWriteInstance_isOwnerAndMember(t *testing.T) {
 	var userId uint64 = 123
-	var groupId uint64 = 321
+	var groupName = "321"
```
Comment (teleivo) [inline internal/handler/authorize_test.go:13]: I would call the var `group`, if there is no name collision.

File: internal/server/engine.go [hunk 3]
```diff
@@ -35,9 +35,9 @@ func GetEngine(environment di.Environment) *gin.Engine {
 	tokenAuthenticationRouter.DELETE("/instances/:id", environment.InstanceHandler.Delete)
 	tokenAuthenticationRouter.GET("/instances/:id", environment.InstanceHandler.FindById)
 	tokenAuthenticationRouter.GET("/instances/:id/logs", environment.InstanceHandler.Logs)
-	tokenAuthenticationRouter.GET("/instances-name-to-id/:groupId/:name", environment.InstanceHandler.NameToId)
+	tokenAuthenticationRouter.GET("/instances-name-to-id/:groupName/:instanceName", environment.InstanceHandler.NameToId)
 
-	tokenAuthenticationRouter.POST("/instances/:id/save", environment.InstanceHandler.Save)
+	//tokenAuthenticationRouter.POST("/instances/:id/save", environment.InstanceHandler.Save)
```
Comment (teleivo) [inline internal/server/engine.go:40]: why is that commented out?

File: pkg/instance/handler.go [hunk 4]
```diff
@@ -506,9 +505,9 @@ func (h Handler) Logs(c *gin.Context) {
 }
 
 // NameToId instance
-// swagger:route GET /instances-name-to-id{groupId}/{name} instanceNameToId
+// swagger:route GET /instances-name-to-id{groupName}/{instanceName} instanceNameToId
```
Comment (teleivo) [inline pkg/instance/handler.go:508]: I think there is a `/` missing before `groupName`

File: pkg/instance/repository.go [hunk 5]
```diff
@@ -48,10 +46,13 @@ func (r repository) FindWithParametersById(id uint) (*model.Instance, error) {
 	return instance, err
 }
 
-func (r repository) FindByNameAndGroup(instanceName string, groupId uint) (*model.Instance, error) {
+func (r repository) FindByNameAndGroup(instanceName string, groupName string) (*model.Instance, error) {
```
Comment (teleivo) [inline pkg/instance/repository.go:49]: If there is no name collision I would name the args `instance` and `group`. The method says FindByName. Name is also now the identifier for almost all things. In an ambiguous context we can postfix the `id` or `name`. Its clear here

File: pkg/instance/service.go [hunk 6]
```diff
@@ -214,17 +214,17 @@ func (s service) FindWithDecryptedParametersById(id uint) (*model.Instance, erro
 	return instance, nil
 }
 
-func (s service) FindByNameAndGroup(instanceName string, groupId uint) (*model.Instance, error) {
-	return s.instanceRepository.FindByNameAndGroup(instanceName, groupId)
+func (s service) FindByNameAndGroup(instanceName string, groupName string) (*model.Instance, error) {
```
Comment (teleivo) [inline pkg/instance/service.go:217]: same as in the repo

File: pkg/model/stack.go [hunk 7]
```diff
@@ -1,22 +1,30 @@
...
-	Instances          []Instance               `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
+	Name               string                   `gorm:"primaryKey" json:"name"`
+	RequiredParameters []StackRequiredParameter `gorm:"many2many:required_stack_parameters_joins; foreignKey:Name; References:Name; joinForeignKey:StackName; joinReferences:ParameterID; constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"requiredParameters"`
+	OptionalParameters []StackOptionalParameter `gorm:"many2many:optional_stack_parameters_joins; foreignKey:Name; References:Name; joinForeignKey:StackName; joinReferences:ParameterID; constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"optionalParameters"`
+	Instances          []Instance               `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;" json:"instances"`
+
+	//	gorm.Model
+	CreatedAt time.Time
+	UpdatedAt time.Time
+	DeletedAt gorm.DeletedAt `gorm:"index"`
```
Comment (teleivo) [inline pkg/model/stack.go:18]: why did you add these fields + index on DeletedAt?

File: pkg/model/stack.go [hunk 8]
```diff
@@ -1,22 +1,30 @@
...
-	OptionalParameters []StackOptionalParameter `gorm:"many2many:optional_stack_parameters_joins; foreignKey:ID; joinForeignKey:StackID; References:Name; joinReferences:ParameterID; constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"optionalParameters,omitempty"`
-	Instances          []Instance               `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
+	Name               string                   `gorm:"primaryKey" json:"name"`
+	RequiredParameters []StackRequiredParameter `gorm:"many2many:required_stack_parameters_joins; foreignKey:Name; References:Name; joinForeignKey:StackName; joinReferences:ParameterID; constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"requiredParameters"`
+	OptionalParameters []StackOptionalParameter `gorm:"many2many:optional_stack_parameters_joins; foreignKey:Name; References:Name; joinForeignKey:StackName; joinReferences:ParameterID; constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"optionalParameters"`
+	Instances          []Instance               `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;" json:"instances"`
+
+	//	gorm.Model
```
Comment (teleivo) [inline pkg/model/stack.go:15]: why is that commented out?

File: scripts/dhis2-deploy.sh [hunk 9]
```diff
@@ -8,18 +8,18 @@ IMAGE_REPOSITORY=core
 IMAGE_TAG=2.36.0-tomcat-8.5.34-jre8-alpine
 DATABASE_SIZE=30Gi
 PGADMIN_INSTALL=false
-DATABASE_ID=3
```
Comment (teleivo) [inline scripts/dhis2-deploy.sh:11]: why did you change the DB id?

**Review by teleivo — COMMENTED** [review]
added some questions/suggestions

**Review by radnov — COMMENTED** [review]
Took a quick peek - seems alright to me, as long as it works. :shipit:
